### PR TITLE
Add UUCP locking.

### LINKED ISSUE #888: stageprofi plugin disrupting usbpro plugin via /dev/ttyUSB0
When both the usbpro and the stageprofi plugins are available and enabled, the stageprofi plugin will disrupt a functioning usbpro ttyUSB0 and cause it to hang until olad is restarted. Here is an snippet of a log from olad (-l 3) wherein I launch with all plugins enabled, wait for my usbpro to become active, and start sending RDM GET queries via ola_rdm_get.py. You can see that everything is working until I try to transmit while the stageprofi has temporary and erroneous ownership of the ttyUSB0, after which I get the enttec watchdog trigger and perpetual hang.

If I build OLA with support for only the usbpro plugin, or if I specifically disable the stageprofi plugin via its config file, the issue is resolved.

```
plugins/usbpro/UsbProWidgetDetector.cpp:517: Detected USB Device: ESTA Id: 0x0, device Id: 0, serial: 0x2168947, f/w version: 2.4
plugins/usbpro/WidgetDetectorThread.cpp:326: Defaulting to a Usb Pro device
olad/plugin_api/DeviceManager.cpp:105: Installed device: Enttec Usb Pro Device:5-02168947
plugins/usbpro/EnttecUsbProWidget.cpp:276: Incremental discovery triggered
plugins/usbpro/EnttecUsbProWidget.cpp:308: Un-muting all devices, TN: 0
olad/plugin_api/PortManager.cpp:119: Patched 5-02168947-O-0 to universe 1
plugins/usbpro/EnttecUsbProWidget.cpp:308: Un-muting all devices, TN: 1
plugins/usbpro/EnttecUsbProWidget.cpp:308: Un-muting all devices, TN: 2
plugins/usbpro/EnttecUsbProWidget.cpp:331: Sending DUB packet: 0000:00000000 - ffff:ffffffff
common/rdm/DiscoveryAgent.cpp:217: BranchComplete, got 24
common/rdm/DiscoveryAgent.cpp:319: Muting 3638:12000001
plugins/usbpro/EnttecUsbProWidget.cpp:290: Muting 3638:12000001, TN: 4
plugins/usbpro/EnttecUsbProWidget.cpp:475: Probably muted device
plugins/usbpro/EnttecUsbProWidget.cpp:331: Sending DUB packet: 0000:00000000 - ffff:ffffffff
common/rdm/DiscoveryAgent.cpp:217: BranchComplete, got 0
plugins/stageprofi/StageProfiPlugin.cpp:121: New StageProfiWidget: /dev/ttyUSB0
olad/plugin_api/DeviceManager.cpp:105: Installed device: StageProfi Device:8-/dev/ttyUSB0
plugins/stageprofi/StageProfiWidget.cpp:125: No response from StageProfiWidget
plugins/stageprofi/StageProfiPlugin.cpp:147: StageProfi device /dev/ttyUSB0 was removed
plugins/stageprofi/StageProfiPlugin.cpp:121: New StageProfiWidget: /dev/ttyUSB0
olad/plugin_api/DeviceManager.cpp:105: Installed device: StageProfi Device:8-/dev/ttyUSB0
plugins/stageprofi/StageProfiWidget.cpp:125: No response from StageProfiWidget
plugins/stageprofi/StageProfiPlugin.cpp:147: StageProfi device /dev/ttyUSB0 was removed
plugins/stageprofi/StageProfiPlugin.cpp:121: New StageProfiWidget: /dev/ttyUSB0
olad/plugin_api/DeviceManager.cpp:105: Installed device: StageProfi Device:8-/dev/ttyUSB0
plugins/stageprofi/StageProfiWidget.cpp:125: No response from StageProfiWidget
plugins/stageprofi/StageProfiPlugin.cpp:147: StageProfi device /dev/ttyUSB0 was removed
olad/plugin_api/Universe.cpp:459: Universe 1, RDM request to 3638:12000001, SD: 2, CC 0x20, TN 0, PID 0x0060, PDL: 0
olad/plugin_api/Universe.cpp:459: Universe 1, RDM request to 3638:12000001, SD: 2, CC 0x20, TN 0, PID 0x0060, PDL: 0
olad/plugin_api/Universe.cpp:459: Universe 1, RDM request to 3638:12000001, SD: 2, CC 0x20, TN 0, PID 0x0060, PDL: 0
olad/plugin_api/Universe.cpp:459: Universe 1, RDM request to 3638:12000001, SD: 2, CC 0x20, TN 0, PID 0x0060, PDL: 0
olad/plugin_api/Universe.cpp:459: Universe 1, RDM request to 3638:12000001, SD: 2, CC 0x20, TN 0, PID 0x0060, PDL: 0
plugins/stageprofi/StageProfiPlugin.cpp:121: New StageProfiWidget: /dev/ttyUSB0
olad/plugin_api/DeviceManager.cpp:105: Installed device: StageProfi Device:8-/dev/ttyUSB0
olad/plugin_api/Universe.cpp:459: Universe 1, RDM request to 3638:12000001, SD: 2, CC 0x20, TN 0, PID 0x0060, PDL: 0
plugins/usbpro/EnttecUsbProWidget.cpp:612: Enttec watchdog fired
```

Here is my ola-usbserial.conf:
```
device_dir = /dev
device_prefix = ttyUSB0
enabled = true
pro_fps_limit = 190
tri_use_raw_rdm = false
ultra_fps_limit = 40
uucp_lock_path = /var/lock
uucp_lock_path = /tmp
```

### PR DESCRIPTION
**Pull request by nomis52**
Closes #888

### CODE REVIEW

File: common/io/Serial.cpp [hunk 1]
```diff
@@ -31,6 +46,59 @@ namespace io {
 using std::vector;
 using std::string;
 
+namespace {
+
+string GetLockFile(const string &path) {
+  const string base_name = ola::file::FilenameFromPath(path);
```
Comment (peternewman) [inline common/io/Serial.cpp]: I think we had this discussion before, but this potentially falls down if some files aren't in /dev.

File: common/io/Serial.cpp [hunk 2]
```diff
@@ -31,6 +46,59 @@ namespace io {
...
+bool GetPidFromFile(const string &lock_file, pid_t *pid) {
+  int fd = open(lock_file.c_str(), O_RDONLY);
+  if (fd < 0) {
+    if (errno == ENOENT) {
+      // If it doesn't exist we're ok, any other error should be treated as if
+      // a lock exists.
+      *pid = 0;
+      return true;
+    }
+    OLA_INFO << "Failed to read pid from " << lock_file << ": "
```
Comment (peternewman) [inline common/io/Serial.cpp]: SPaG PID

File: common/io/Serial.cpp [hunk 3]
```diff
@@ -31,6 +46,59 @@ namespace io {
...
+
+  char buffer[100];
+  int r = read(fd, buffer, arraysize(buffer));
+  close(fd);
+  if (r < 0) {
+    return false;
+  }
+
+  if (!StringToInt(string(buffer, r), pid)) {
+    OLA_DEBUG << "Failed to convert contents of " << lock_file;
```
Comment (peternewman) [inline common/io/Serial.cpp]: Is it worth printing the string here, or would that be a security risks?

File: common/io/Serial.cpp [hunk 4]
```diff
@@ -55,17 +123,81 @@ bool UIntToSpeedT(uint32_t value, speed_t *output) {
...
+
+  if (locked_pid) {
+    // This will return false even if we have the lock, this is what we want
+    // since different plugins may try to open the same serial port - see issue
+    // #888.
+    if (ProcessExists(locked_pid)) {
+      return false;
+    }
+    // There is a race between the read & the unlink here. I'm not convinced it
+    // can be solved.
```
Comment (peternewman) [inline common/io/Serial.cpp]: Variable delay perhaps? Probably not ideal though.

File: common/io/Serial.cpp [hunk 5]
```diff
@@ -55,17 +123,81 @@ bool UIntToSpeedT(uint32_t value, speed_t *output) {
...
+  pid_t our_pid = getpid();
+  // Now try to open the file in O_EXCL mode. If we fail, someone else already
+  // took the lock.
+  int lock_fd = open(lock_file.c_str(), O_RDWR | O_CREAT | O_EXCL,
+                     S_IRUSR | S_IWUSR | S_IRGRP | S_IWGRP | S_IROTH | S_IWOTH);
+  if (lock_fd < 0) {
+    return false;
+  }
+
+  OLA_INFO << "Acquired " << lock_file;
```
Comment (peternewman) [inline common/io/Serial.cpp]: Acquired lock

File: common/io/Serial.cpp [hunk 6]
```diff
@@ -55,17 +123,81 @@ bool UIntToSpeedT(uint32_t value, speed_t *output) {
...
+
+  // write our pid to the file
```
Comment (peternewman) [inline common/io/Serial.cpp]: SPaG PID

File: common/io/Serial.cpp [hunk 7]
```diff
@@ -55,17 +123,81 @@ bool UIntToSpeedT(uint32_t value, speed_t *output) {
...
+
+  pid_t locked_pid;
+  if (!GetPidFromFile(lock_file, &locked_pid)) {
+    return;
+  }
+
+  pid_t our_pid = getpid();
+  if (our_pid == locked_pid) {
+    if (RemoveLockFile(lock_file)) {
+      OLA_INFO << "Released " << lock_file;
```
Comment (peternewman) [inline common/io/Serial.cpp]: Released lock

**Comment by peternewman** [comment]
Is it worth adding the ability to override the output from configure as a command line argument?

File: common/io/Serial.cpp (hunk 1 above)
Comment (nomis52) [inline common/io/Serial.cpp]: Yup

File: common/io/Serial.cpp (hunk 2 above)
Comment (nomis52) [inline common/io/Serial.cpp]: Done

File: common/io/Serial.cpp (hunk 3 above)
Comment (nomis52) [inline common/io/Serial.cpp]: The contents could be binary so I don't want to blindly print it.

File: common/io/Serial.cpp (hunk 4 above)
Comment (nomis52) [inline common/io/Serial.cpp]: That's not going to help, in fact it'll make the problem worse.

File: common/io/Serial.cpp (hunk 5 above)
Comment (nomis52) [inline common/io/Serial.cpp]: the name contains LCK, so it's clear it's a lock file. I'm trying to reduce logging output

File: common/io/Serial.cpp (hunk 7 above)
Comment (nomis52) [inline common/io/Serial.cpp]: Same

**Comment by nomis52** [comment]
Use   --with-uucp-lock    to override the lock directory.

File: common/io/Serial.cpp (hunk 3 above)
Comment (peternewman) [inline common/io/Serial.cpp]: Okay.

File: common/io/Serial.cpp (hunk 4 above)
Comment (peternewman) [inline common/io/Serial.cpp]: I was thinking two copies of OLA getting run up at the same time, but I can see the issue with two plugins now too.

File: common/io/Serial.cpp (hunk 5 above)
Comment (peternewman) [inline common/io/Serial.cpp]: Okay, I'd have thought line count is the thing that really matters, rather than characters.

**Comment by peternewman** [comment]
I meant as an option to olad @nomis52 , to avoid having to recompile.

**Comment by peternewman** [comment]
LGTM, just one outstanding comment.

**Comment by nomis52** [comment]
Maybe one day when we have global options or something. This is not likely to change (ever)

**Comment by peternewman** [comment]
Wouldn't it just be a case of doing something similar to what we do for --http-data-dir ?

**Pull request merged**